qt/python/mantidqt/project/recovery/projectrecovery.py

```diff
@@ -0,0 +1,471 @@
+            logger.debug("Project Recovery: Failed to save error msg: " + str(e))
+
+        # Spin off another timer thread
+        if not self.closing_workbench:
+            self._spin_off_another_time_thread()
+
+    def _spin_off_another_time_thread(self):
+        self._timer_thread = Timer(self.time_between_saves, self.recovery_save)
+        self._timer_thread.start()
+
+    def _save_workspaces(self, directory):
+        # Get all present workspaces
+        ws_list = ADS.getObjectNames()
+
+        if len(ws_list) == 0:
+            return
+
+        start_time = UsageService.getStartTime().toISO8601String()
+
+        alg_name = "GeneratePythonScript"
+        alg = AlgorithmManager.createUnmanaged(alg_name, 1)
+        alg.setChild(True)
+        alg.setLogging(False)
+
+        for index, ws in enumerate(ws_list):
+            if self._empty_group_workspace(ws):
+                continue
+
+            filename = str(index) + ".py"
+            filename = os.path.join(directory, filename)
+
+            alg.initialize()
+            alg.setProperty("AppendTimestamp", True)
+            alg.setProperty("InputWorkspace", ws)
+            alg.setPropertyValue("Filename", filename)
+            alg.setPropertyValue("StartTimestamp", start_time)
+            alg.setProperty("IgnoreTheseAlgs", self.algs_to_ignore)
+            alg.setProperty("IgnoreTheseAlgProperties", self.alg_properties_to_ignore)
+
+            alg.execute()
+
+    @staticmethod
+    def _empty_group_workspace(ws):
+        if isinstance(ws, WorkspaceGroup) and len(ws.getNames()) == 0:
+            return True
+
+    def _save_project(self, directory, interfaces_list=None):
+        project_saver = ProjectSaver(project_file_ext=self.recovery_file_ext)
+
+        plots = self.gfm.figs
+        if interfaces_list is None:
+            interfaces_list = self.interface_finding_func()
+
+        project_saver.save_project(directory, workspace_to_save=None, plots_to_save=plots,
+                                   interfaces_to_save=interfaces_list, save_workspaces=False)
+
+    def _add_lock_file(self, directory):
+        # Create the file
+        open(os.path.join(directory, self.lock_file_name), 'a').close()
+
+    def _remove_lock_file(self, directory):
+        lock_file = os.path.join(directory, self.lock_file_name)
+        if os.path.exists(lock_file):
+            os.remove(lock_file)
+
+    ######################################################
+    #  Decision
+    ######################################################
+
+    def check_for_recover_checkpoint(self):
+        try:
+            # Clean directory first
+            self._remove_empty_folders_from_dir(self.recovery_directory_hostname)
+
+            checkpoints = self.listdir_fullpath(self.recovery_directory_hostname)
+            num_of_mantids = self._number_of_other_workbench_processes()
+            return len(checkpoints) != 0 and len(checkpoints) > num_of_mantids
```

Comment:
I'm not quite sure I understand how the number of checkpoints relate to the number of mantids running?

Reply:
One PID-(checkpoint) equals one mantid process. If the number of checkpoints is less than the number of (other) mantid processes then a recovery can be expected.

This should be cleared with better variable names and a comment. I will add these shortly.